### ISSUE TITLE
Design Eval feature for scripts

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#5985]
It should be easy to evaluate dynamically created source code from within a script:

my_script.csx:

```C#
var result = await Script.Eval("1+1");

var result = await Script.Eval("X+Y", new C { X = 1, Y = 2});
```

The provided capabilities would be similar to the Scripting APIs.

Some questions:
- Can we pass in info from the enclosing scope (such as usings)?
  Using CallerFileName and CallerLineNumber attributes the host, which implements the Eval method, can get the exact line in the source code of the Eval call. From there it can infer the usings in scope, if it has the compilation still available. If the script was executed via CreateDelegate the compilation might not be available anymore (by design, to free up resources). That can be ok, we might just not support evals in this scenario and throw an exception.

**Comment by tmat** [comment]
@mattwar @MadsTorgersen @dotnet/interactive

**Comment by glennblock** [comment]
+1 I have had several use cases where I would like to do this. It would be nice to support #r, using, #load and have this become a submission in the main session.

**Comment by mattwar** [comment]
Seems like a lot of magic to have it infer usings, though really useful.  

Might be interesting to have default globals object be the current execution state, thus capturing all declared variables so far.

**Comment by tmat** [comment]
@mattwar Yes, eval in scripting languages is usually full of magic :) Many times too much (Ruby, Perl). I think we can do with just a little bit of magic.

**Comment by glennblock** [comment]
Yes, we introduced a ton of our own magic with scriptcs. Magic in scripting is part of the allure :-)

**Comment by tmat** [comment]
@glennblock Well, there is magic and there is magic ;)

**Comment by glennblock** [comment]
Fair that we use some more potent ingredients :-)
